### ISSUE TITLE
replace klog.Fatal and os.Exit for defer funcs

### ISSUE THREAD

**Issue by sanposhiho** [kubernetes/kubernetes#102231]
Hello, team 👋

#### What happened:

The defer function will not be called if os.Exit(1).
https://play.golang.org/p/f7IjXeb62dd

> deferred functions are not run.
https://golang.org/pkg/os/#Exit

Now, the entry points of some Kubernetes components use defer functions to flush logs.

For example, kube-scheduler
https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-scheduler/scheduler.go#L41
```
func main() {
	rand.Seed(time.Now().UnixNano())

	pflag.CommandLine.SetNormalizeFunc(cliflag.WordSepNormalizeFunc)

	command := app.NewSchedulerCommand()

	logs.InitLogs()
	defer logs.FlushLogs()

	if err := command.Execute(); err != nil {
		os.Exit(1)
	}
}
```

When `command.Execute()` returns an error, logs are not flushed. Is this an expected behavior?

#### How to fix

We can ensure that `defer` is executed by separating the functions as follows.

```
func main() {
  if err := realmain(); err != nil {
    os.Exit(1)
  }
}

func realmain() error {
	rand.Seed(time.Now().UnixNano())

	pflag.CommandLine.SetNormalizeFunc(cliflag.WordSepNormalizeFunc)

	command := app.NewSchedulerCommand()

	logs.InitLogs()
	defer logs.FlushLogs()

	if err := command.Execute(); err != nil {
		return err
	}
       return nil
}
```

can I pick this issue? 
/assign

## Edit (October 23)

For entry points, https://github.com/kubernetes/kubernetes/pull/105076 resolves the issue. 
But it is better to delete all klog.Fatal and os.Exit as much as possible. 
They can be replaced by returning error/nil. And in some case, we may be able to use panic() (defer funcs are executed when panic [ref](https://go.dev/blog/defer-panic-and-recover))

**Comment by logicalhan** [comment]
I think we ran into something like this in apiserver, which was the reason for this particular health check: https://github.com/kubernetes/kubernetes/blob/b2b4cb17256de8d89d4d5fdbe8d19a2492e8152c/staging/src/k8s.io/apiserver/pkg/server/healthz/healthz.go#L69-L83

**Comment by ash2k** [comment]
I'd like to add that it's not only logs. I've seen some kubectl tests that use commands and then when there is an error, such test terminates the whole test process rather than a particular test, which is annoying. FWIW I support the idea of fixing this repo-wide. I suggest to identify all the components where this issue applies and send an email to corresponding SIGs to get their approval on this before doing any work.

**Comment by sanposhiho** [comment]
Hello, @ash2k 
> I suggest to identify all the components where this issue applies 

Okay.
>  and send an email to corresponding SIGs to get their approval on this before doing any work.

Is it good to create issues for each component instead of email? It is easy to follow the latest status from here.

**Comment by ash2k** [comment]
I think one issue (this one) should be enough to start the conversation.

**Comment by sanposhiho** [comment]
>  one issue (this one) should be enough to start the conversation.

OK

---

I checked all entrypoints under /cmd, and these have this issue.

- cloud-controller-manager (sig/api-machinery, sig/cloud-provider)
- kube-apiserver (sig/api-machinery)
- kube-controller-manager (sig/api-machinery)
- kube-proxy (sig/network)
- kube-scheduler (sig/scheduling) (I've already created a PR for scheduler. #104503 )
- kubectl (sig/cli)
- kubectl-convert  (sig/cli)
- kubelet (sig/node)
- kubemark (sig/scalability)

I'll check to corresponding SIGs on each email list.

**Comment by kerthcet** [comment]
hi @sanposhiho , I can provide some help if you needs. 🧐

**Comment by fedebongio** [comment]
thank you Walter for taking this one!
/triage accepted

**Comment by kerthcet** [comment]
should we push these related prs forward?🎉

**Comment by kerthcet** [comment]
I have checked #104774, #104775, #104776 already, the only left one #104773 I'll working on it this week.

**Comment by ash2k** [comment]
This is great. Happy to see all those `os.Exit()` calls removed.

There is at least one more thing that calls `os.Exit()` - https://github.com/kubernetes/kubernetes/blob/dac94e1c9ee31034d174815e287ba39fa67b2b72/staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go#L114-L116. It's used in a lot of places, such as https://github.com/kubernetes/kubernetes/blob/dac94e1c9ee31034d174815e287ba39fa67b2b72/staging/src/k8s.io/kubectl/pkg/cmd/logs/logs.go#L157-L159.

**Comment by sanposhiho** [comment]
@kerthcet 

By #105076, we don't have to create another func for entry points.
So, we only have to delete os.Exit on cobra commands for this issue :)

**Comment by cheftako** [comment]
The lease/leader election use of os.Exit is deliberate and has been baked into several controllers. We can't remove it till we have an alternate way to ensure we don't have competing controllers.

**Comment by sanposhiho** [comment]
> The lease/leader election use of os.Exit is deliberate and has been baked into several controllers. We can't remove it till we have an alternate way to ensure we don't have competing controllers.

Yes. I think there are places where `os.Exit` is definitely needed. If we were to introduce linter, we would need to use nolint comment-directive there.
https://golangci-lint.run/usage/false-positives/#nolint-directive

It is difficult to make all developers aware of the need to avoid using `os.Exit`, but, linter can do. Hopefully, developers will avoid using `os.Exit` as much as possible and use nolint only when necessary.

**Comment by sanposhiho** [comment]
Okay, I just found that the latest forbidigo provides the useful feature; custom forbid reason messages.
https://github.com/ashanbrown/forbidigo/pull/11

So, if golangci-lint updates its dependencies, we can use this feature and show error message like: 
`Please avoid to use os.Exit as much as possible. see #102231 for more details. If you still think we need to use os.Exit here, you can use "//nolint forbidigo // reason" here to avoid this lint issue.`

**Comment by sanposhiho** [comment]
So, to sum up, I would propose this migration plan:
- find where `os.Exit` or `klog.Fatal` is used and replace them with appropriate way. (returning error, using panic, just log error or ... etc)
  - For places where they are definitely needed, add `//nolint:forbidigo // reason` to avoid `forbidigo` linter which will be enabled in the next step.
- After replacing, enable `forbidigo` linter and configure it to forbid `os.Exit` and `klog.Fatal`.
  - also configure custom forbid reason messages; https://github.com/kubernetes/kubernetes/issues/102231#issuecomment-1019274061

What do you all think about this?

**Comment by logicalhan** [comment]
> This will be related to `klog`. (Sorry if it's not appropriate sig)
> 
> /sig instrumentation

It's a usage (or mis-use) of klog, so I'm not sure we're actually the right SIG for this.

**Comment by felipeagger** [comment]
I would like to contribute to this issue.
/assign

**Comment by adrianmoisey** [comment]
/unassign 

(I'm not sure I understand the scope of this issue. I'm happy to pick it up again if someone can help me understand what needs fixing)
